### ISSUE TITLE
[SDK] Allow user to select WCAG AA or AAA, the light and dark text colors and light text opacity for the Design System.  Those are the basis for all contrast calcualtions

### ISSUE THREAD

**Issue by lwnoble** [finos/a11y-theme-builder#746]
## Problem/Concern

Allow user to select wcag AA or AAA, text colors and dark mode light text opacity levels

## Proposed Solution

This is dependant on the changes made to in issue 501

We introduce these new variables that can be set by the user:
   var wcagContrast 
   var wcagSMContrast (which should forever be 3.1)
   var darkText 
   var lightText
   var dmWhiteOpacity

I updated the following functions to include these variables

1.) checkContrast
2.) adjustLighterCount
3.) adjustLighter
4.) adjustDarkerCount
5.) adjustDarker

    function checkContrast(theme, color, mode) {
      var lightTextArray = hextoRGBArray(lightText);
      var darkTextArray  = hextoRGBArray(darkText)
      var rgbArray = hextoRGBArray(rgb2hex(color));
      var shade = theme.split('-')[2];
      var newRGB = "rgb(" + rgbArray +")"
      if (mode == 'dark') {
        var colorHex = rgb2hex(color)
        var newHex = mixColors(colorHex,lightText,dmWhiteOpacity );
        var lightArray = hex2rgb(newHex)
        var lightArray = lightArray.replace('rgb(','')
        var lightArray = lightArray.replace(')','');
        var lightArray = lightArray.split(',');
      } else {
        var lightArray = lightTextArray
      }
      console.log('shade: ' + shade + ', lightArray: ' + lightArray)
      var light = contrast(lightArray, rgbArray);
      var dark  = contrast(darkTextArray, rgbArray);
      var text_color, textTint, contrastRation
      if (wcagContrast == 7.1) {
        if (shade <= 500) {
          text_color = darkTextArray; // dark
          var textTint = 'dark';
          contrastRation = contrast(text_color, rgbArray);
        } else {
          text_color = lightArray; // white
          var textTint = 'light';
          contrastRation = contrast(lightArray, rgbArray);
        }
      } else {
        if ( light > dark ) {
          text_color = lightArray; // white
          var textTint = 'light';
          contrastRation = contrast(lightArray, rgbArray);
        } else {
          text_color = darkTextArray; // dark
          var textTint = 'dark';
          contrastRation = contrast(text_color, rgbArray);
        }
      }

      if (textTint == 'light') {
        var buildText = lightTextArray
      } else {
        var buildText =  darkTextArray
      }

      contrastRation  = contrastRation.toFixed(2)
      if (contrastRation <= wcagContrast) {
        var darkCount      = adjustDarkerCount(theme, newRGB, lightArray, contrastRation, mode)
        var lightCount     = adjustLighterCount(theme, newRGB, darkTextArray, contrastRation, mode)
        if (darkCount < lightCount || shade >= 600) {
          adjustColorDarker(theme, newRGB, lightArray, contrastRation, mode)
        } else {
          adjustColorLighter(theme, newRGB, darkTextArray, contrastRation, mode)
        }
      } else {
        console.log('theme: ' + theme + ' ,text color:' + text_color + ', rgbArray' +  rgbArray  + ', contrastRation: ' + contrastRation)
        buildColor(theme,mode,rgbArray,buildText,contrastRation)
      }
    }

// make color .01 lighter //
  function adjustLighterCount(theme, color, text_color, contrastRation, mode) {
    var i = 0
    var multiplier = .01
    if (color.indexOf("rgb") >= 0) {
      color = rgb2hex(color)
    }
    while (contrastRation < wcagContrast) {
      var rgbArray = [];
      var r = hexToRgb(color).r;
      var g = hexToRgb(color).g;
      var b = hexToRgb(color).b;
      var newR = Math. round(r + ((250 - r) * multiplier ))
      if (newR == r) {
        newR = r + 1
      }
      if (newR > 255) {
        newR = 255
      }
      var newG = Math. round(g + ((250 - g) * multiplier ))
      if (newG == g) {
        newG = g + 1
      }
      if (newG > 255) {
        newG = 255
      }
      var newB = Math. round(b + ((250 - b) * multiplier ))
      if (newB == b) {
        newB = b + 1
      }
      if (newB > 255) {
        newB = 255
      }
      rgbArray.push(newR);
      rgbArray.push(newG);
      rgbArray.push(newB);
      var rgb = 'rgb('+rgbArray+')'
      var hex = rgb2hex(rgb)
      var contrastRation = contrast(rgbArray, text_color); // 1.0736196319018405
      var backgroundColor = $('.background:not(.darkmode)').css('backgroundColor')
      var contrastToBG = contrast(rgbArray, text_color); // 1.0736196319018405
      if (contrastRation >= wcagContrast){
        return(i)
      }
      color = rgb2hex(rgb)
    }
  }

     // make color .01 lighter //
     function adjustColorLighter(theme, color, text_color, contrastRation, mode) {
       var multiplier = .01
       if (color.indexOf("rgb") >= 0) {
         color = rgb2hex(color)
       }
       while (contrastRation < wcagContrast) {
         var rgbArray = [];
         var r = hexToRgb(color).r;
         var g = hexToRgb(color).g;
         var b = hexToRgb(color).b;
         var newR = Math. round(r + ((250 - r) * multiplier ))
         if (newR == r) {
           newR = r + 1
         }
         if (newR > 255) {
           newR = 255
         }
         var newG = Math. round(g + ((250 - g) * multiplier ))
         if (newG == g) {
           newG = g + 1
         }
         if (newG > 255) {
           newG = 255
         }
         var newB = Math. round(b + ((250 - b) * multiplier ))
         if (newB == b) {
           newB = b + 1
         }
         if (newB > 255) {
           newB = 255
         }
         rgbArray.push(newR);
         rgbArray.push(newG);
         rgbArray.push(newB);
         var rgb = 'rgb('+rgbArray+')'
         var hex = rgb2hex(rgb)
         var contrastRation = contrast(rgbArray, text_color); // 1.0736196319018405
         var backgroundColor = $('.background:not(.darkmode)').css('backgroundColor')
         var contrastToBG = contrast(rgbArray, text_color); // 1.0736196319018405
         console.log(i + ' Lighten:' + theme + ', ' + color + ', ' + text_color + ', ' + contrastRation + ', ' + mode)

         if (contrastRation >= wcagContrast){
           contrastRation  = contrastRation.toFixed(2)
           color = rgbArray
           buildColor(theme, mode,rgbArray,text_color,contrastRation)
           $('#' + theme ).addClass('lightened')
           break;
         }
         color = rgb2hex(rgb)
       }
     }

     // create color .01 darker //
     function adjustDarkerCount(theme, color, text_color, contrastRation, mode) {
       console.log('adjust: '+ color + ', text_color: '+ text_color)
       var i = 0
       if (color.indexOf("rgb") >= 0) {
         color = rgb2hex(color)
       }
       var multiplier = .99
       while (contrastRation < wcagContrast) {
         i = i + 1
         var rgbArray = [];
         var r = hexToRgb(color).r;
         var g = hexToRgb(color).g;
         var b = hexToRgb(color).b;
         var newR = Math.round(r * multiplier)
         if (newR == r ) {
           newR = r - 1
         }
         var newG = Math.round(g * multiplier)
         if (newG == g ) {
           newG = g - 1
         }
         var newB = Math.round(b * multiplier)
         if (newB == b ) {
           newB = b - 1
         }
         if (newR <= 0) {
            newR = 0
          }
          if (newG <= 0) {
            newG = 0
          }
          if (newB <= 0) {
            newB = 0
          }
         rgbArray.push(newR);
         rgbArray.push(newG);
         rgbArray.push(newB);
         console.log('g: ' + g + ', newG: ' + newG)
         var rgb = 'rgb('+rgbArray+')'
         if (mode == 'dark' && text_color == [255,255,255]) {
           var hex = rgb2hex(rgb)
           var colorHex = rgb2hex(color)
           var newHex = mixColors(colorHex,'#FFFFFF',dmWhiteOpacity );
           var lightArray = hex2rgb(newHex)
           var lightArray = lightArray.replace('rgb(','')
           var lightArray = lightArray.replace(')','');
           text_color = lightArray.split(',');
         }
         var contrastRation = contrast(rgbArray, text_color);
         var contrastRation = contrastRation.toFixed(2)
         if (contrastRation >= wcagContrast){
          return(i)
         }
         color = rgb2hex(rgb)
         console.log('rgb:'  + rgb2hex(rgb) + ', contrast:' + contrastRation)
       }
     }

     // create color .01 darker //
     function adjustColorDarker(theme, color, text_color, contrastRation, mode) {
       i = 0
       if (color.indexOf("rgb") >= 0) {
         color = rgb2hex(color)
       }
       var multiplier = .99
       while (contrastRation < wcagContrast) {
         var rgbArray = [];
         var r = hexToRgb(color).r;
         var g = hexToRgb(color).g;
         var b = hexToRgb(color).b;

         var newR = Math. round(r * multiplier)
         if (newR == r ) {
           newR = r - 1
         }
         var newG = Math. round(g * multiplier)
         if (newG == g ) {
           newG = g - 1
         }
         var newB = Math. round(b * multiplier)
         if (newB == b ) {
           newB = b - 1
         }
         if (newR <= 0) {
            newR = 0
          }
          if (newG <= 0) {
            newG = 0
          }
          if (newB <= 0) {
            newB = 0
          }
         rgbArray.push(newR);
         rgbArray.push(newG);
         rgbArray.push(newB);
         var rgb = 'rgb('+rgbArray+')'
         var hex = rgb2hex(rgb)
         if (mode == 'dark' && text_color == '255,255,255') {
           var colorHex = rgb2hex(color)
           var newHex = mixColors(colorHex,'#FFFFFF',dmWhiteOpacity );
           var lightArray = hex2rgb(newHex)
           var lightArray = lightArray.replace('rgb(','')
           var lightArray = lightArray.replace(')','');
           var lightArray = lightArray.split(',');
           var contrastRation = contrast(lightArray, text_color);
           var contrastRation = contrastRation.toFixed(2)
         } else {
           var contrastRation = contrast(rgbArray, text_color); // 1.0736196319018405
           contrastRation  = contrastRation.toFixed(2)
         }
         if (contrastRation >= wcagContrast){
          $('#' + theme ).addClass('darkened')
           console.log('Darkened: ' +  mode + ', ' + theme + ', ' + color + ', ' + ', ' + contrastRation)
           buildColor(theme, mode,rgbArray,[255,255,255],contrastRation);
           break;
         }
         i = i++
         color = rgb2hex(rgb)
       }
     }

![Uploading Screenshot 2023-12-12 at 11.04.30 am.png…]()

Please see my zip file demo.

**Comment by lwnoble** [comment]
Please note the paragraph spacing should be updated as follows:

AA = 2 x the font size
AAA = 2 x the line height

The min-target will also need to be automatically updated to a min of 44px if the user selects AAA.
